app/src/org/commcare/activities/RecoveryActivity.java

```diff
@@ -37,259 +37,241 @@
         }
         return network;
     }
 
-    private void attemptRecovery() {
-        AndroidCommCarePlatform platform = CommCareApplication.instance().getCommCarePlatform();
+    public void attemptRecovery() {
+        appManagerBt.setVisibility(View.INVISIBLE);
+        loadingIndicator.setVisibility(View.VISIBLE);
+
+        CommCareApplication commCareApplication = CommCareApplication.instance();
+        // Try to reinitialize to refresh the list of missing resources
+        commCareApplication.initializeAppResources(commCareApplication.getCurrentApp());
+        AndroidCommCarePlatform platform = commCareApplication.getCommCarePlatform();
         ResourceTable global = platform.getGlobalResourceTable();
         if (!global.getMissingResources().isEmpty()) {
-            int total = global.getMissingResources().size();
-            ResourceRecoveryTask<RecoveryActivity> task =
-                    new ResourceRecoveryTask<RecoveryActivity>(RECOVERY_TASK) {
-
-                        @Override
-                        protected void deliverResult(RecoveryActivity recoveryActivity, Boolean success) {
-                            if (success) {
-                                // Try Reinitializing and attemptRecovery again to correct against further issues
-                                CommCareApplication.instance().initializeAppResources(CommCareApplication.instance().getCurrentApp());
-                                attemptRecovery();
-                            } else {
-                                recoveryActivity.displayMessage(StringUtils.getStringRobust(recoveryActivity, R.string.recovery_error));
-                            }
-                        }
-
-                        @Override
-                        protected void deliverUpdate(RecoveryActivity recoveryActivity, Integer... update) {
-                            int done = update[0];
-                            recoveryActivity.updateProgress(
-                                    StringUtils.getStringRobust(recoveryActivity, R.string.recovery_resource_progress,
-                                            new String[]{String.valueOf(done), String.valueOf(total)}),
-                                    RECOVERY_TASK);
-                            updateProgressBar(done, total, RECOVERY_TASK);
-                        }
-
-                        @Override
-                        protected void deliverError(RecoveryActivity recoveryActivity, Exception e) {
-                            Logger.exception("Error while recovering missing resources " + ForceCloseLogger.getStackTrace(e), e);
-                            recoveryActivity.displayMessage(StringUtils.getStringRobust(recoveryActivity, R.string.recovery_error));
-                        }
-                    };
+            ResourceRecoveryTask task = ResourceRecoveryTask.getInstance();
             task.connect(this);
             task.executeParallel();

```

Comment:
Does executeParallel crash out if the task is already running and/or handle it gracefully? 

Crashing out is still better than letting two execute at the same time, but flagging that if the getInstance returns an already existing instance it'll cause issues here.

I think the "Safest" pattern would be something like

```
task = ResourceRecoveryTask.getInstance();
task.connect()
if(taskstate = unstarted) { task.execute}

ResourceRecoveryTask.getInstanceAndConnect(connector) {
    lock {
          //since the lock code right now happens on the UI thread, the task can actually finish before it's disconnected
          if  ( exists.state is cancelled or complete) { exists = null}
          if  ( exists = null) { new exists() }
          return exists;
    }
}
```

does that make sense?

Reply:
Updated the `getNewInstance` to use the throw exception pattern Philip used for `UpdateTask`

Reply:
That looks good to me, thanks for jumping back into that